test/tap-filter

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+#
+# Tap parser that filters out test success and only prints the errors and
+# test summary.
+#
+# The parser keeps a buffer of all output during a test (including non-tap
+# data). If that test has a failing assertion, the entire buffer is printed.
+#
+# This filter adds one non-standard tap output line:
+#
+#    #: This is an example line that will be printed immediately
+#
+# You can use the "#:" prefix to a line and it will be passed through the
+# filter to stdout. This is using for printing status information.
+#
+declare test_name=
+declare test_error=false
+declare buffer=
+declare exit_code=0
+
+#
+# Add the passed-in data to the buffer.
+#
+_add_to_buffer() {
+	for data in "$@"; do
+		buffer="$buffer"$'\n'"$data"
+	done
+}
+
+while IFS= read -r line; do
+	case $line in
+		\#*)
+			# Beginning of new test...
+
+			# If there was an error during the last test, print the error
+			# before resetting
+			if $test_error; then
+				echo >&2
```

Comment:
cool, I would do this for sure if there was anything else being printed on stderr elsewhere in the script. As it stands, creating a function would mean the reader has to go to another place to trace what is happening.

One other possible way to write it would also be:

```
				{
					echo
					echo -e "\033[1;31m${test_name}\033[0m"
					echo "$buffer"
				} >&2
```

I think I'll just leave it as-is, though.